src/providers/oracle/qgsoracleprovider.cpp

```diff
@@ -2505,6 +2506,118 @@ bool QgsOracleProvider::setSubsetString( const QString &theSQL, bool updateFeatu
+    return result;
+  }
+
+  QString sql(
+    "SELECT c.constraint_name, "
+    "     l.column_name, "
+    "     r.owner, "
+    "     r.table_name, "
+    "     r.column_name "
+    "FROM all_constraints c, "
+    "     all_cons_columns l, "
+    "     all_cons_columns r "
+    "WHERE l.table_name = c.table_name "
+    "  AND c.table_name = ?"
+    "  AND c.owner = ?"
+    "  AND c.constraint_type = 'R'"
+    "  AND c.r_constraint_name = r.constraint_name "
+    "  AND c.constraint_name = l.constraint_name "
+    "  AND l.position = r.position "
+  );
+  QgsOracleConn *conn = connectionRO();
+  QSqlQuery qry( *conn );
+  int nbFound = 0;
+  QList<QString> refTableFound;
+  if ( !LoggedExecStatic( qry, sql, QVariantList() << mTableName << mOwnerName, mUri.uri() ) )
+  {
+    const QString error { tr( "Unable to execute the query to get foreign keys of %1.\nThe error message from the database was:\n%1.\nSQL: %2" )
+                          .arg( mTableName )
+                          .arg( qry.lastError().text() )
+                          .arg( qry.lastQuery() ) };
+    QgsLogger::warning( error );
+    return result;
+  }
+
+  while ( qry.next() )
+  {
+    const QString name = qry.value( 0 ).toString();
+    const QString fkColumn = qry.value( 1 ).toString();
+    const QString refOwner = qry.value( 2 ).toString();
+    const QString refTable = qry.value( 3 ).toString();
+    const QString refColumn = qry.value( 4 ).toString();
+
+    if ( ( nbFound == 0 ) || ( !refTableFound.contains( refTable ) ) )
+    {
+      // first reference field => try to find if we have layers for the referenced table
+      const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refOwner, refTable );
+      for ( const QgsVectorLayer *foundLayer : foundLayers )
+      {
+        QgsRelation relation;
+        relation.setName( name );
+        relation.setReferencingLayer( target->id() );
+        relation.setReferencedLayer( foundLayer->id() );
+        relation.addFieldPair( fkColumn, refColumn );
+        relation.generateId();
+        if ( relation.isValid() )
+        {
+          result.append( relation );
+          ++nbFound;
+          refTableFound.append( refTable );
+        }
+        else
+        {
+          QgsLogger::warning( "Invalid relation for " + name + ", validation error: " + relation.validationError() );
+        }
+      }
+    }
+    else
+    {
+      // multi reference field => add the field pair to all the referenced layers found
+      const QList<QgsVectorLayer *> foundLayers = searchLayers( layers, mUri.connectionInfo( false ), refOwner, refTable );
+      for ( int i = 0; i < nbFound; ++i )
+      {
+        for ( const QgsVectorLayer *foundLayer : foundLayers )
+        {
+          if ( result[result.size() - 1 - i].referencedLayerId() == foundLayer->id() )
+          {
+            result[result.size() - 1 - i].addFieldPair( fkColumn, refColumn );
+          }
```

Comment:
else what? If we have a multi ref field to a table existing in several layers, it wouldn't work correctly no?